Sources/AsyncKit/ConnectionPool/EventLoopGroupConnectionPool.swift

```diff
@@ -178,14 +185,110 @@ public final class EventLoopGroupConnectionPool<Source> where Source: Connection
         }
     }
     
-    deinit {
-        // synchronize access to didShutdown
-        self.lock.lock()
-        defer { self.lock.unlock() }
+    /// Closes the connection pool.
+    ///
+    /// All available connections will be closed immediately. Any connections still in use will be
+    /// closed as soon as they are returned to the pool. Once closed, the pool can not be used to
+    /// create new connections.
+    ///
+    /// Connection pools must be closed before they deinitialize.
+    ///
+    /// This method shuts down synchronously, waiting for all connection closures to complete before
+    /// returning.
+    ///
+    /// - Warning: The pool is always fully shut down once this method returns, even if an error is
+    ///   thrown. All errors are purely advisory.
+    public func syncShutdownGracefully() throws {
+        // - TODO: Does this need to assert "not on any EventLoop", as `EventLoopGroup.syncShutdownGracefully()` does?
```

Comment:
I think this would be good to add.

Reply:
Well, the thrust of the question in the comment was less "is it nice to check as good behavior" (which would be at  most an assert) and more "does this algorithm in fact require that precondition in order to function safely?" (a precondition)

Reply:
Hmm that's a good question. I'd probably prefer an error over deadlock, even in production code, so that makes me think `precondition`. What does NIO do?

Reply:
I wasn't actually clear on whether it really did have that requirement to begin with, to be sure... Thinking it through, for a `pool.close()` to complete properly, its `EventLoop` can't be blocked, which means the calling code in synchronous mode can't be on that loop, which in turn mean it can't be on _any_ of the event loops in the group.

That leaves the question of do we try to assert "not on any event loop in the group" (which is complex) or just "not on any kind of event loop at all" (which is the way in which NIO just skips to the end and one-liner's it)? The Fluent drivers aren't going to be able to switch to async shutdown without breaking API, so it'd be nice to not limit them more than really necessary... I think the specific group can be asserted with something like `precondition(self.eventLoopGroup.makeIterator().forEach { $0.preconditionNotInEventLoop() })` (requires NIO 2.17.0, which merged my PR with the "not in" variant).